### ISSUE TITLE
terraform + helm + multiple CRDs

### ISSUE THREAD

**Issue by AndyMoore** [Azure/aad-pod-identity#161]
I'm trying to use aad-pod-identity for multiple pods giving them access to different resources (e.g. app1 will access keyvault, external-dns will modify dns zones). This would be fine - aiui I just need to create the 2 CRDs and update the pod labels accordingly.

The problem I've run into is that I'm trying to do things in a 100% automated way, but terraform isnt currently able to setup arbitrary CRD resources.. 

The only solution in my mind (which I agree is the wrong place, and should be done in terraform) would be to update the aad-pod-identity helm chart in one of 2 ways:

- put a flag around creating the CRDs so I can have multiple deployments of aad-pod-identity each looking after their own pod and identity/role definitions, but only one copy of the CRDs (https://github.com/Azure/aad-pod-identity/blob/master/charts/aad-pod-identity/templates/crds.yaml)
- investigate allowing the helm chart to setup multiple identities (via https://github.com/Azure/aad-pod-identity/blob/master/charts/aad-pod-identity/templates/identities.yaml)

So first question - is there a better way, and if not would either of the above make sense, and (lastly) if none of the above are there any other options? :) 

I'd be happy to make the changes and return them as a PR, otherwise I'll probably have to clone this and put it into its own pipeline within our own CI which I'm not overly keen to do either

**Comment by kkmsft** [comment]
cc @jjcollinge @ritazh

**Comment by jjcollinge** [comment]
@AndyMoore - appreciate this is less than ideal situation. I've hit this limitation in TF a number of times and had to rely on the Helm provider to get around it. My colleague @lawrencegripper wrote a [PR to add CRD support](https://github.com/terraform-providers/terraform-provider-kubernetes/pull/195) to the Kubernetes provider but as you can see from the thread, Hashi are waiting on a few things before enabling this behaviour. He did, however, pull this out into a stand-alone provider which is available here: https://github.com/lawrencegripper/terraform-provider-kubernetes-yaml
Unfortunately, I think it's a case of pick your poison with this one. If you don't want to use a fork of the kubernetes TF provider or a stand-alone provider, then forking this Helm chart and adding your conditional logic maybe the best solution - if you're happy to keep it up to date with the upstream. I'll defer to @kkmsft and @ritazh to whether they'd accept that addition as a PR into the upstream or not.

**Comment by lawrencegripper** [comment]
On [my Terraform YAML support fork there is still outstanding work to get CRD behaving correctly](https://github.com/lawrencegripper/terraform-provider-kubernetes-yaml/issues/11), it's on my todo list but currently other projects taking all my time.

**Comment by AndyMoore** [comment]
thanks both. i've gone down the road of creating new charts for each service (e.g. external-dns) that i'm deploying that'd use aad-pod-identity, and including the CRDs in there. that's not ideal because it'll fail should aad-pod-identity not be deployed in time (which still creates the CRDs). I went down that road because aad-pod-identity took me a _lot_ longer to configure and get working than the others, so in retrospect I'd rather use upstream

I did see @lawrencegripper 's provider, but the maturity is low where I'm at, so I really don't want to introduce new dependencies and manual steps where I can avoid them. I'm going to have enough other issues with the relative immaturity of the azure provider within terraform already :)

None of this is ideal though, so I might reassess at some point in the not too distant future. I think to sum up I don't like any of the approaches, and am desperate for someone to come along with something I haven't seen before..

**Comment by lawrencegripper** [comment]
@AndyMoore one which I've been hoping to play with is [weaveworks flux](https://github.com/weaveworks/flux) as the approach looks nice but as I've not used it yet not sure now/whether it would fix this scenario and it also introduces another thing to learn

**Comment by AndyMoore** [comment]
I've looked at that briefly too, more specifically https://github.com/weaveworks/flux/blob/master/site/helm-operator.md - the problem here is that it uses CRDs.. 😃 

I appreciate this is kinda off topic for this project, and specifically for the issues page (and my apologies), but I'm now looking for ways to continuously deploy apps to an ephemeral (i.e. one that can be stood up and destroyed by an engineer on demand) AKS cluster that doesnt include any more terraform than is needed.. terraform for the infrastructure, but not for what goes onto kubernetes I think

**Comment by kkmsft** [comment]
@AndyMoore - Thank you for this issue and discussions. Do we have any action specific to Aad-pod-identity project here ?

**Comment by AndyMoore** [comment]
maybe just documentation for the scenario for now. worth mentioning, but I'm not sure any real actions for this project

**Comment by kkmsft** [comment]
@AndyMoore - Thank you! Would you be interested in raising a PR for documenting this ?

**Comment by AndyMoore** [comment]
I would, but I'm still not much the wiser for the 'proper' solution :) 

helm + crds seems to be a grey area

**Comment by aramase** [comment]
Duplicate of #285. I've a PR in the official charts repo with steps on manually cleaning up CRD after removing the charts. Please follow #285 for further discussion.
